### PR TITLE
allow underscore in number literals

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 1]
```diff
@@ -1782,12 +1782,15 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
     case Some(t) => WeededAst.Expression.Ascribe(exp0, Some(visitType(t)), None, exp0.loc)
   }
 
+  private def stripSeps(digits: String): String = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala]: Add documentation. Rename to "stripUnderscores".

**Comment by magnus-madsen** [comment]
I have been carefully reading the Java spec which states:


You can place underscores only between digits; you cannot place underscores in the following places:

- At the beginning or end of a number.
- Adjacent to a decimal point in a floating point literal.
- Prior to an F or L suffix.
- In positions where a string of digits is expected.

I would like to see at least these two enforced:

- At the beginning or end of a number.
- Adjacent to a decimal point in a floating point literal.

File: main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Int.flix [hunk 2]
```diff
@@ -25,3 +25,6 @@ def testInt07(): Int = 2147483647
 
 @test
 def testInt08(): Int = -2147483648
+
+@test
+def testInt09(): Int = 1_0
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/feature/Test.Expression.Int.flix]: I'd like to see a handful of test cases for each type. For Int32, for example, I would expect to see:

```
1_000
42_000
1_000_000
42_000_000
1_000_000_000
42_000_000_000
```

or the like. Just to make sure we cover more ground.

**Comment by mlutze** [comment]
The change that I made corresponds to the regex [`\d(?:_*\d)*`](https://regex101.com/r/FDkqvm/1), which does fulfill all requirements there when used within the individual literal parsers. I opted not to create an alternate rule since that would have left the `Digits` rule unused, but I'll separate them for clarity.

I think this might have been clearer with some negative tests. Those would need to go into a new ParserTest Suite since we can't test failure to compile in *.flix tests.

**Comment by magnus-madsen** [comment]
> The change that I made corresponds to the regex [`\d(?:_*\d)*`](https://regex101.com/r/FDkqvm/1), which does fulfill all requirements there when used within the individual literal parsers. I opted not to create an alternate rule since that would have left the `Digits` rule unused, but I'll separate them for clarity.
> 
> I think this might have been clearer with some negative tests. Those would need to go into a new ParserTest Suite since we can't test failure to compile in *.flix tests.

Sorry. You are absolutely right. I see it now. I must have been tired :)

It would be good with a few more positive and negative test cases and then this is ready.

**Comment by mlutze** [comment]
I added more test cases. Notably missing from test cases are those where the digits begin with underscore (e.g. `_100`). This is because in those cases, the string is still valid and parsable, but as an identifier rather than a literal, so a ParseError is not appropriate there.

**Comment by magnus-madsen** [comment]
This looks good to me! Thank you. I am ready to merge it if you say it is ready.

**Comment by mlutze** [comment]
It's ready!

**Pull request merged**